### PR TITLE
access_log: formatter for upstream host metadata

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
formatter for upstream host metadata

Additional Description:
Part of #20389. Most of the formatters used in header manipulation are also present in substitution access log formatters. However, UPSTREAM_METADATA was not present in access log formatters.

Also, as noted in #17457 all xxxx_METADATA will be eventually replaced my METADATA(xxxx,...) so this PR also extends METADATA formatter.

api changes are trivial and limited to comments.
Risk Level: Low
Testing: Added unit tests.
Docs Changes: Yes.
Release Notes: Yes
Platform Specific Features: No

### CODE REVIEW

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by ggreenway** [comment]
@dmitri-d please review.

**Comment by cpakulski** [comment]
@dmitri-d Would you mind reviewing the changes? They are really trivial. The formatter just needs to get the metadata from HostDescription, similarly to getting dynamic and cluster matadata as it is done here: https://github.com/envoyproxy/envoy/blob/release/v1.22/source/common/formatter/substitution_formatter.cc#L1683-L1702.

Thanks!

File: source/common/formatter/substitution_formatter.cc [hunk 1]
```diff
@@ -1806,6 +1816,23 @@ ClusterMetadataFormatter::ClusterMetadataFormatter(const std::string& filter_nam
...
+                                                             absl::optional<size_t> max_length)
+    : MetadataFormatter(filter_namespace, path, max_length,
+                        [](const StreamInfo::StreamInfo& stream_info)
+                            -> const envoy::config::core::v3::Metadata* {
+                          if (!stream_info.upstreamInfo().has_value()) {
+                            return nullptr;
+                          }
+                          Upstream::HostDescriptionConstSharedPtr host =
+                              stream_info.upstreamInfo()->upstreamHost();
+                          if (!host) {
```
Comment (lizan) [inline source/common/formatter/substitution_formatter.cc:1830]: nit:
```suggestion
                          if (host != nullptr) {
```

**Review by lizan — COMMENTED** [review]
LGTM other than one nit

File: source/common/formatter/substitution_formatter.cc (hunk 1 above)
Comment (cpakulski) [reply source/common/formatter/substitution_formatter.cc]: Thanks for reviewing. Corrected.

**Review by lizan — APPROVED** [review]

**Comment by lizan** [comment]
@htuch PTAL for non-tetrands review

**Review by htuch — APPROVED** [review]

**Pull request merged**